rllib/algorithms/algorithm.py

```diff
@@ -941,6 +943,199 @@ def duration_fn(num_units_done):
+
+            _agent_steps = sum(b.agent_steps() for b in batches)
+            _env_steps = sum(b.env_steps() for b in batches)
+
+            # 1 episode per returned batch.
+            if unit == "episodes":
+                num_units_done += len(batches)
+                # Make sure all batches are exactly one episode.
+                for ma_batch in batches:
+                    ma_batch = ma_batch.as_multi_agent()
+                    for batch in ma_batch.policy_batches.values():
+                        assert np.sum(batch[SampleBatch.DONES])
+            # n timesteps per returned batch.
+            else:
+                num_units_done += (
+                    _agent_steps if self._by_agent_steps else _env_steps
+                )
+            if self.reward_estimators:
+                all_batches.extend(batches)
+
+            agent_steps_this_iter += _agent_steps
+            env_steps_this_iter += _env_steps
+
+            logger.info(
+                f"Ran round {round_} of parallel evaluation "
+                f"({num_units_done}/{duration if not auto else '?'} "
+                f"{unit} done)"
+            )
+
+        if timed_out and log_once("evaluation_timeout"):
+            logger.warning(
+                "Calling `sample()` on your remote evaluation worker(s) "
+                "resulted in a timeout (after the configured "
+                f"{self.config['evaluation_sample_timeout_s']} seconds)! "
+                "Try to set `evaluation_sample_timeout_s` in your config"
+                " to a larger value."
+                + (
+                    " If your episodes don't terminate easily, you may "
+                    "also want to set `evaluation_duration_unit` to "
+                    "'timesteps' (instead of 'episodes')."
+                    if unit == "episodes"
+                    else ""
+                )
+            )
+
+        metrics = summarize_episodes(
+            rollout_metrics,
+            keep_custom_metrics=eval_cfg["keep_per_episode_custom_metrics"],
+        )
+
+        metrics[NUM_AGENT_STEPS_SAMPLED_THIS_ITER] = agent_steps_this_iter
+        metrics[NUM_ENV_STEPS_SAMPLED_THIS_ITER] = env_steps_this_iter
+        # TODO: Remove this key at some point. Here for backward compatibility.
+        metrics["timesteps_this_iter"] = env_steps_this_iter
+
+        if self.reward_estimators:
+            # Compute off-policy estimates
+            metrics["off_policy_estimator"] = {}
+            total_batch = concat_samples(all_batches)
+            for name, estimator in self.reward_estimators.items():
+                estimates = estimator.estimate(total_batch)
+                metrics["off_policy_estimator"][name] = estimates
+
+        # Evaluation does not run for every step.
+        # Save evaluation metrics on trainer, so it can be attached to
+        # subsequent step results as latest evaluation result.
+        self.evaluation_metrics = {"evaluation": metrics}
```

Comment:
let's move this block into _run_one_evaluation()? then we don't have to duplicate this in both evaluate_v2() and evaluate()

Reply:
I think this won't work. What if the user wants to call `evaluate` manually?